libs/wingc/src/lsp/sync.rs

```diff
@@ -1,42 +1,59 @@
-use lsp_types::{DidChangeTextDocumentParams, DidOpenTextDocumentParams, Url};
+use indexmap::IndexMap;
+use lsp_types::{DidChangeTextDocumentParams, DidOpenTextDocumentParams};
 use wingii::type_system::TypeSystem;
 
-use std::path::Path;
-use std::{cell::RefCell, collections::HashMap};
+use std::cell::RefCell;
+use std::path::{Path, PathBuf};
 use tree_sitter::Tree;
 
 use crate::closure_transform::ClosureTransformer;
-use crate::diagnostic::{get_diagnostics, reset_diagnostics, Diagnostic};
+use crate::diagnostic::{found_errors, reset_diagnostics, Diagnostic};
 use crate::files::Files;
 use crate::fold::Fold;
 use crate::jsify::JSifier;
-use crate::parser::Parser;
+use crate::lifting::LiftTransform;
+use crate::parser::{parse_wing_project, FileGraph};
 use crate::type_check;
 use crate::type_check::jsii_importer::JsiiImportSpec;
+use crate::type_check_assert::TypeCheckAssert;
 use crate::{ast::Scope, type_check::Types, wasm_util::ptr_to_string};
 
-/// The result of running wingc on a file
-pub struct FileData {
-	/// Text data contained in the file (ut ≥f8)
-	pub contents: String,
-	/// tree-sitter tree
-	pub tree: Tree,
+/// The output of compiling a Wing project with one or more files
+pub struct ProjectData {
+	/// Text data contained in the files (utf8)
+	pub files: Files,
+	/// A graph that tracks the dependencies between files
+	pub file_graph: FileGraph,
+	/// tree-sitter trees
+	pub trees: IndexMap<PathBuf, Tree>,
 	/// The diagnostics returned by wingc
 	pub diagnostics: Vec<Diagnostic>,
-	/// The top scope of the file
-	pub scope: Box<Scope>,
-	/// The universal type collection for the scope. This is saved to ensure references live long enough.
-	pub types: Types,
-	/// The JSII imports for the file. This is saved so we can load JSII types (for autotocompletion for example)
+	/// AST for each file
+	pub asts: IndexMap<PathBuf, Scope>,
+	/// The JSII imports for the file. This is saved so we can load JSII types (for autocompletion for example)
 	/// which don't exist explicitly in the source.
 	pub jsii_imports: Vec<JsiiImportSpec>,
 }
 
+impl ProjectData {
+	fn new() -> Self {
+		ProjectData {
+			files: Files::new(),
+			file_graph: FileGraph::default(),
+			trees: IndexMap::new(),
+			diagnostics: Vec::new(),
+			asts: IndexMap::new(),
+			jsii_imports: Vec::new(),
+		}
+	}
+}
+
 thread_local! {
 	/// When consumed as a WASM library, wingc is not in control of the process/memory in which it is running.
 	/// This means that it cannot reliably manage stateful data like this between function calls.
 	/// Here we will assume the process is single threaded, and use thread_local to store this data.
-	pub static FILES: RefCell<HashMap<Url,FileData>> = RefCell::new(HashMap::new());
+	pub static WING_TYPES: RefCell<Types> = RefCell::new(Types::new());
+	pub static PROJECT_DATA: RefCell<ProjectData> = RefCell::new(ProjectData::new());
```

Comment:
I was running into some hairy borrow checking errors (due to trying to mutably borrow `Types` and immutably borrow an AST `Scope` at the same time) at some point - the best solution I could come up with was to put "Types" into its own thread-local variable.